test/ha/ha.go

```diff
@@ -0,0 +1,157 @@
+/*
+Copyright 2020 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package ha
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"net/url"
+	"strings"
+	"testing"
+	"time"
+
+	appsv1 "k8s.io/api/apps/v1"
+	autoscalingv1 "k8s.io/api/autoscaling/v1"
+	corev1 "k8s.io/api/core/v1"
+	apierrs "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	restclient "k8s.io/client-go/rest"
+	pkgTest "knative.dev/pkg/test"
+	rtesting "knative.dev/serving/pkg/testing/v1"
+	"knative.dev/serving/test"
+	"knative.dev/serving/test/e2e"
+	v1test "knative.dev/serving/test/v1"
+)
+
+const (
+	servingNamespace = "knative-serving"
+	haReplicas       = 2
+)
+
+func getLeader(t *testing.T, clients *test.Clients, labelSelector string) (string, error) {
+	var leader string
```

Comment:
is there no other way besides logs parsing?

Reply:
Another way is to wait for events of type Lease. Does it sound better? If yes, why?

Reply:
Waiting on logs sounds a bit hacky to me. If we have a real event to watch (that is Lease updates), I'd prefer that.

Reply:
Can't we watch the configMap the thing takes its lease on?

Reply:
I've change the impl to poll the Lease object. The config-leader-election configmap doesn't give me the name of the leader pod. It doesn't have useful information and we use it only to enable/disable the leader election.

Reply:
Yeah I meant the Lease object.